src/tools/r2rdump/SignatureType.cs

```diff
@@ -38,20 +43,26 @@ public SignatureType(ref BlobReader signatureReader, MetadataReader mdReader, Ge
         {
             SignatureTypeCode signatureTypeCode = signatureReader.ReadSignatureTypeCode();
             Flags = 0;
-            if (signatureTypeCode == SignatureTypeCode.SZArray)
+
+            if (signatureTypeCode == SignatureTypeCode.ByReference)
             {
-                Flags |= SignatureTypeFlags.ARRAY;
+                Flags |= SignatureTypeFlags.REFERENCE;
                 signatureTypeCode = signatureReader.ReadSignatureTypeCode();
             }
-
-            TypeName = signatureTypeCode.ToString();
-            if (signatureTypeCode == SignatureTypeCode.TypeHandle || signatureTypeCode == SignatureTypeCode.ByReference)
+            if (signatureTypeCode == SignatureTypeCode.SZArray)
             {
-                if (signatureTypeCode == SignatureTypeCode.ByReference)
+                Flags |= SignatureTypeFlags.ARRAY;
+                ArrayDimension = 0;
+                while (signatureTypeCode == SignatureTypeCode.SZArray)
                 {
-                    Flags |= SignatureTypeFlags.REFERENCE;
+                    signatureTypeCode = signatureReader.ReadSignatureTypeCode();
```

Comment:
This looks potentially dangerous because reading from the signature reader will increment its offset, and it may affect future uses of the `signatureReader` object that is passed into this method. I recommend saving the Offset value before the read and then resetting it when exiting the while loop.

Reply:
The part that uses signatureReader after the while loop needs the offset to read what comes after the SZArray typecodes. Should it reset the offset at the end of this function instead?

Reply:
I guess I'm curious what happens when you finish reading the SZArray's. It seems like it may fail the next read because you have already read the bytes directly after the SZArray type code. Does it actually fail or does this work correctly right now?

Reply:
It works correctly right now with arrays of objects or generic types when I tried it. The typecodes for object arrays are [SZArray]...[SZArray][TypeHandle][Handle] and it reads the first typecode at the beginning of the function, so whether it's an array or not, [TypeHandle] would already be read after the while loop

Reply:
Can you address Jan's comment about jagged arrays here? Is there any work left to be done on that?

Reply:
I've made some changes for getting the signature so that it uses the existing MethodDefinition.DecodeSignature instead of my own. That would handle jagged arrays and some other cases